crypto/fipsmodule/ec/p384.c

```diff
@@ -0,0 +1,434 @@
+
+  fiat_p384_square(x3, x2);   // 2^3 - 2^1
+  fiat_p384_mul(x3, x3, in);  // 2^3 - 2^0
+
+  fiat_p384_square(x6, x3);
+  for (int i = 1; i < 3; i++) {
+    fiat_p384_square(x6, x6);
+  }                           // 2^6 - 2^3
+  fiat_p384_mul(x6, x6, x3);  // 2^6 - 2^0
+
+  fiat_p384_square(x12, x6);
+  for (int i = 1; i < 6; i++) {
+    fiat_p384_square(x12, x12);
+  }                             // 2^12 - 2^6
+  fiat_p384_mul(x12, x12, x6);  // 2^12 - 2^0
+
+  fiat_p384_square(x15, x12);
+  for (int i = 1; i < 3; i++) {
+    fiat_p384_square(x15, x15);
+  }                             // 2^15 - 2^3
+  fiat_p384_mul(x15, x15, x3);  // 2^15 - 2^0
+
+  fiat_p384_square(x30, x15);
+  for (int i = 1; i < 15; i++) {
+    fiat_p384_square(x30, x30);
+  }                              // 2^30 - 2^15
+  fiat_p384_mul(x30, x30, x15);  // 2^30 - 2^0
+
+  fiat_p384_square(x60, x30);
+  for (int i = 1; i < 30; i++) {
+    fiat_p384_square(x60, x60);
+  }                              // 2^60 - 2^30
+  fiat_p384_mul(x60, x60, x30);  // 2^60 - 2^0
+
+  fiat_p384_square(x120, x60);
+  for (int i = 1; i < 60; i++) {
+    fiat_p384_square(x120, x120);
+  }                                // 2^120 - 2^60
+  fiat_p384_mul(x120, x120, x60);  // 2^120 - 2^0
+
+  fiat_p384_felem ret;
+  fiat_p384_square(ret, x120);
+  for (int i = 1; i < 120; i++) {
+    fiat_p384_square(ret, ret);
+  }                                // 2^240 - 2^120
+  fiat_p384_mul(ret, ret, x120);   // 2^240 - 2^0
+
+  for (int i = 0; i < 15; i++) {
+    fiat_p384_square(ret, ret);
+  }                                // 2^255 - 2^15
+  fiat_p384_mul(ret, ret, x15);    // 2^255 - 2^0
+
+  for (int i = 0; i < (1 + 30); i++) {
+    fiat_p384_square(ret, ret);
+  }                                // 2^286 - 2^31
+  fiat_p384_mul(ret, ret, x30);    // 2^286 - 2^30 - 2^0
+
+  fiat_p384_square(ret, ret);
+  fiat_p384_square(ret, ret);      // 2^288 - 2^32 - 2^2
+  fiat_p384_mul(ret, ret, x2);     // 2^288 - 2^32 - 2^0
+
+  for (int i = 0; i < (64 + 30); i++) {
+    fiat_p384_square(ret, ret);
+  }                                // 2^382 - 2^126 - 2^94
+  fiat_p384_mul(ret, ret, x30);    // 2^382 - 2^126 - 2^94 + 2^30 - 2^0
+
+  fiat_p384_square(ret, ret);
+  fiat_p384_square(out, ret);      // 2^384 - 2^128 - 2^96 + 2^32 - 2^2 = p - 3
+}
+
+// Group operations
+// ----------------
+//
+// Building on top of the field operations we have the operations on the
+// elliptic curve group itself. Points on the curve are represented in Jacobian
+// coordinates.
+//
+
+// The method is taken from:
+//   http://hyperelliptic.org/EFD/g1p/auto-shortw-jacobian-3.html#doubling-dbl-2001-b
+//
+// Coq transcription and correctness proof:
+// <https://github.com/mit-plv/fiat-crypto/blob/79f8b5f39ed609339f0233098dee1a3c4e6b3080/src/Curves/Weierstrass/Jacobian.v#L93>
+// <https://github.com/mit-plv/fiat-crypto/blob/79f8b5f39ed609339f0233098dee1a3c4e6b3080/src/Curves/Weierstrass/Jacobian.v#L201>
```

Comment:
The P-256 implementation states that x_out can be the same as x_in, etc. I assume it's the same here?

Reply:
I think it is actually, because if you look in `ec_GFp_nistp256_point_mul_base()` for example, the `fiat_p384_point_double()` is called with the same input and output arguments.

Reply:
Thanks for this observation, Dusan.